File: Dynamic_Programming/132.Palindrome-Partitioning-II/132.Palindrome-Partitioning-II.py
class Solution:
    def minCut(self, s):
        """
        :type s: str
        :rtype: int
        """
        N = len(s)
        isPal = [[0 for _ in range(N)] for _ in range(N)]
        for i in range(N): isPal[i][i]=1
        for i in range(N-1): 
            if s[i]==s[i+1]: isPal[i][i+1]=1 
        for length in range(3,N+1):
            for start in range(N-length+1):
                if s[start]==s[start+length-1] and isPal[start+1][start+length-2]==1:
                    isPal[start][start+length-1]=1
                
        dp = list(range(1,N+1))        
        for i in range(1,N):
            if isPal[0][i]==1:
                dp[i]=1
            else:
                for j in range(i):
                    if isPal[j+1][i]==1:
                        dp[i]=min(dp[i],dp[j]+1)
        
        return dp[N-1]-1

File: Dynamic_Programming/132.Palindrome-Partitioning-II/132.Palindrome-Partitioning-II.cpp
class Solution {
public:
    int minCut(string s) 
    {
        int N = s.size();
        
        auto isPal = vector<vector<bool>>(N,vector<bool>(N,0));
        for (int len = 1; len<=N; len++)
            for (int i=0; i<=N-len; i++)
            {
                int j = i+len-1;
                // update isPal[i][j]
                if (s[i]==s[j])
                {
                    if (i+1>=j-1)
                        isPal[i][j] = true;
                    else 
                        isPal[i][j] = isPal[i+1][j-1];
                }
            }
        
        
        vector<int>dp(N, INT_MAX/2);
        
        dp[0] = 1;
        
        for (int i=1; i<N; i++)
        {
            for (int j=0; j<=i; j++)
            {
                if (isPal[j][i])
                {
                    if (j==0)
                        dp[i] = 1;
                    else
                        dp[i] = min(dp[i], dp[j-1]+1);
                }
            }
        }
        
        return dp[N-1]-1;
    }
    
};
